applications/plugins/SofaMiscCollision/src/SofaMiscCollision/DefaultCollisionGroupManager.cpp

```diff
@@ -285,8 +304,132 @@ simulation::Node* DefaultCollisionGroupManager::getIntegrationNode(core::Collisi
+    }
+
+    if (!node->linearSolver.empty())
+    {
+        solverSet.linearSolver = *node->linearSolver.begin();
+    }
+
+    if (!node->constraintSolver.empty())
+    {
+        solverSet.constraintSolver = *node->constraintSolver.begin();
+    }
+}
+
+void DefaultCollisionGroupManager::removeSolverSetFromNode(simulation::Node::SPtr node, sofa::component::collision::SolverSet& solverSet)
+{
+    if (solverSet.odeSolver)
+    {
+        node->removeObject(solverSet.odeSolver);
+    }
+    if (solverSet.linearSolver)
+    {
+        node->removeObject(solverSet.linearSolver);
+    }
+    if (solverSet.constraintSolver)
+    {
+        node->removeObject(solverSet.constraintSolver);
+    }
+}
+
+void DefaultCollisionGroupManager::destroySolvers(sofa::component::collision::SolverSet& solverSet)
+{
+    if (solverSet.odeSolver != nullptr)
+    {
+        solverSet.odeSolver.reset();
+    }
+    if (solverSet.linearSolver != nullptr)
+    {
+        solverSet.linearSolver.reset();
+    }
+    if (solverSet.constraintSolver != nullptr)
+    {
+        solverSet.constraintSolver.reset();
+    }
+}
+
+void DefaultCollisionGroupManager::addSolversToNode(simulation::Node::SPtr node, sofa::component::collision::SolverSet& solverSet)
+{
+    if (solverSet.odeSolver)
+    {
+        node->addObject(solverSet.odeSolver);
+    }
+    if (solverSet.linearSolver)
+    {
+        node->addObject(solverSet.linearSolver);
+    }
+    if (solverSet.constraintSolver)
+    {
+        node->addObject(solverSet.constraintSolver);
+    }
+    // perform init only once everyone has been added (in case of explicit dependencies)
+    if (solverSet.odeSolver)
+    {
+        solverSet.odeSolver->init();
+    }
+    if (solverSet.linearSolver)
+    {
+        solverSet.linearSolver->init();
+    }
+    if (solverSet.constraintSolver)
+    {
+        solverSet.constraintSolver->init();
+    }
+}
+
+simulation::Node* DefaultCollisionGroupManager::getNodeFromMergedGroups(simulation::Node* node, const MergeGroupsMap& mergedGroups)
+{
+    if (node != nullptr)
+    {
+        simulation::Node* nodeBegin = node;
+        auto it = mergedGroups.find(node);
 
-} // namespace component
+        //the group associated to a contact may have been moved to another group
+        while (it != mergedGroups.cend() && it->second != nullptr)
+        {
+            node = it->second;
+            checkEndlessLoop(mergedGroups, nodeBegin, node);
+            it = mergedGroups.find(node);
+        }
+    }
+    return node;
+}
+
+bool DefaultCollisionGroupManager::checkEndlessLoop(const MergeGroupsMap& mergedGroups, simulation::Node* firstGroup, simulation::Node* currentGroup)
+{
+    if (currentGroup == firstGroup)
+    {
+        std::stringstream errorStream;
+        for (const auto& [a, b] : mergedGroups)
+        {
+            errorStream << a << " " << b << "\n";
+        }
+        msg_fatal() << "Detected an endless loop:\n" << errorStream.str();
```

Comment:
I'll change the message to the user.
I am surprised `msg_fatal` does not throw an exception or exit the program.
In my case, I call `std::exit`. What do you think?

Reply:
I would throw an exception so that you can easily debug where the error came from the call stack. It should then be the role of the caller (found using the call stack) to prevent calling this method with invalid parameters, and suggesting a warning or error message to the user. If it isn't able to do it, than go up again it the call stack until some method has enough "high level" knowledge to tell the user what he did wrong.